flang-rt/cmake/modules/AddFlangRTOffload.cmake

```diff
@@ -0,0 +1,118 @@
+#===-- cmake/modules/AddFortranRTOffload.txt -------------------------------===#
+#
+# Part of the LLVM Project, under the Apache License v2.0 with LLVM Exceptions.
+# See https://llvm.org/LICENSE.txt for license information.
+# SPDX-License-Identifier: Apache-2.0 WITH LLVM-exception
+#
+#===------------------------------------------------------------------------===#
+
+macro(enable_cuda_compilation name files)
+  if (FLANG_RT_EXPERIMENTAL_OFFLOAD_SUPPORT STREQUAL "CUDA")
+    enable_language(CUDA)
+
+    # TODO: figure out how to make target property CUDA_SEPARABLE_COMPILATION
+    # work, and avoid setting CMAKE_CUDA_SEPARABLE_COMPILATION.
+    set(CMAKE_CUDA_SEPARABLE_COMPILATION ON)
+
+    # Treat all supported sources as CUDA files.
+    set_source_files_properties(${files} PROPERTIES LANGUAGE CUDA)
+    set(CUDA_COMPILE_OPTIONS)
+    if ("${CMAKE_CUDA_COMPILER_ID}" MATCHES "Clang")
+      # Allow varargs.
+      set(CUDA_COMPILE_OPTIONS
+        -Xclang -fcuda-allow-variadic-functions
+        )
+    endif()
+    if ("${CMAKE_CUDA_COMPILER_ID}" MATCHES "NVIDIA")
+      set(CUDA_COMPILE_OPTIONS
+        --expt-relaxed-constexpr
+        # Disable these warnings:
+        #   'long double' is treated as 'double' in device code
+        -Xcudafe --diag_suppress=20208
+        -Xcudafe --display_error_number
+        )
+    endif()
+    set_source_files_properties(${files} PROPERTIES COMPILE_OPTIONS
+      "${CUDA_COMPILE_OPTIONS}"
+      )
+
+    # Create a .a library consisting of CUDA PTX.
+    # This is different from a regular static library. The CUDA_PTX_COMPILATION
+    # property can only be applied to object libraries and create *.ptx files
+    # instead of *.o files. The .a will consist of those *.ptx files only.
+    add_flangrt_library(obj.${name}PTX OBJECT ${files})
+    set_property(TARGET obj.${name}PTX PROPERTY CUDA_PTX_COMPILATION ON)
+    add_flangrt_library(${name}PTX STATIC "$<TARGET_OBJECTS:obj.${name}PTX>")
+
+    # Apply configuration options
+    if (FLANG_RT_CUDA_RUNTIME_PTX_WITHOUT_GLOBAL_VARS)
+      target_compile_definitions(obj.${name}PTX
+        PRIVATE FLANG_RT_NO_GLOBAL_VAR_DEFS
+        )
+    endif()
+
+    # When using libcudacxx headers files, we have to use them
+    # for all files of Flang-RT.
+    if (EXISTS "${FLANG_RT_LIBCUDACXX_PATH}/include")
+      target_include_directories(obj.${name}PTX AFTER PRIVATE "${FLANG_RT_LIBCUDACXX_PATH}/include")
+      target_compile_definitions(obj.${name}PTX PRIVATE RT_USE_LIBCUDACXX=1)
```

Comment:
Also need:
```
      target_include_directories(${name} AFTER PRIVATE "${FLANG_RT_LIBCUDACXX_PATH}/include")
      target_compile_definitions(${name} PRIVATE RT_USE_LIBCUDACXX=1)
```
Because the flang_rt.a library itself is also built with CUDA.

Sorry for not catching it earlier, I missed the build warnings from this and only caught this while doing some end to end.

Reply:
I pushed that changed. However, instead of tons of warnings I now get errors:
```
/home/meinersbur/src/llvm-project-flangrt/flang-rt/../flang/include/flang/Common/variant.h:18:10: fatal error: cuda/std/variant: No such file or directory
   18 | #include <cuda/std/variant>
      |          ^~~~~~~~~~~~~~~~~~
```
CUDA's variant is located in `/usr/include/cuda/std/detail/libcxx/include/variant` (From `libcu++-dev` Ubuntu 24.04.1). There is no package where it is located in `cuda/std/variant`.

Reply:
And you were not getting those when building the PTX lib?
What path did you provide as `FLANG_RT_LIBCUDACXX_PATH`?
Can you check if you have `/usr/include/cuda/std/variant`?

In which case I think you should maybe just provide `/usr` as `FLANG_RT_LIBCUDACXX_PATH`.

@vzakhari, FYI.

Reply:
Yes, `cuda/std/variant` was added after `libcu++` project was converged/replaced with `cccl` (https://github.com/nvidia/cccl).  Here is how the buildbot is configured https://lab.llvm.org/buildbot/#/builders/152/builds/889/steps/4/logs/stdio: `b'-DFLANG_LIBCUDACXX_PATH=/home/buildbot/worker/third-party/nv/cccl/libcudacxx'`

Reply:
I did not set `FLANG_RT_LIBCUDACXX_PATH` at all. I assumed it had found it in a default search path since it compiled without error.

If this requires custom environments I suggest for Nvidia to check and fix it in their own environments. Some documentation would be appreciated as well. I put in the work to keep the old CMake code functional so nothing breaks until `LLVM_ENABLE_RUNTIMES=flang-rt` is added.